### PR TITLE
Type checking with mypy

### PR DESCRIPTION
**Pull request by carmenbianca**
I'd been meaning to do this for a while. We have all these type annotations, but don't check whether they're valid.

### CODE REVIEW

**Comment by carmenbianca** [comment]
Almost completed, except for `report.py` and `lint.py`. I'll wait on #654 to merge before doing that.

**Comment by carmenbianca** [comment]
Not almost completed; I disallowed undefined and incomplete function defs, and will need to do some work on that.

File: src/reuse/_util.py [hunk 1]
```diff
@@ -413,7 +426,12 @@ def _checksum(path: PathLike) -> str:
 class PathType:
     """Factory for creating Paths"""
 
-    def __init__(self, mode="r", force_file=False, force_directory=False):
+    def __init__(
+        self,
+        mode: str = "r",
```
Comment (carmenbianca) [inline src/reuse/_util.py:431]: This type is probably incorrect

Comment (carmenbianca) [reply src/reuse/_util.py:431]: Should be `Literal`, but only available in Python 3.8+.

Comment (nicorikken) [reply src/reuse/_util.py:431]: You could add a comment for future reference:
```suggestion
        mode: str = "r", # change to Literal for python3.8+
```

**Review by nicorikken — APPROVED** [review]
My approval for the current changes, I reviewed all code changes. Thanks, great work!
Besides adding typing, I noticed a couple of improvements in logic, variables and handling like handling copyright statements. In the future perhaps specific types can be used for copyright statements and years as not to rely on `str`. I learned a couple of new constructs.

**Comment by carmenbianca** [comment]
Rebased this on top of #759

File: src/reuse/_util.py (hunk 1 above)
Comment (carmenbianca) [reply src/reuse/_util.py:455]: I looked at this: there are too many possible permutations for me to bother with this. I'll just lazily keep it as-is.

**Pull request merged**